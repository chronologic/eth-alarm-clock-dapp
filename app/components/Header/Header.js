import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { observer,inject } from 'mobx-react';

@inject('web3Service')
@inject('keenStore')
@observer
class Header extends Component {
  constructor(props) {
    super(props);
    this.state = {
      blocknumber: ''
    };
    this.getCurrentBlock = this.getCurrentBlock.bind(this);
  }

  componentWillMount() {
    this.getCurrentBlock();
  }

  componentDidMount() {
    // Check every 10 seconds if the block number changed
    this.interval = setInterval(this.getCurrentBlock, 10000);
  }

  getCurrentBlock() {
    const { web3Service: { web3 } } = this.props;

    web3.eth.getBlockNumber((err,res) =>{
      err == null && this.setState({ blocknumber: res });
    });
  }

  componentWillUnmount() {
    clearInterval(this.interval);
  }

  render() {
    return (
      <div className="header">
        <a href="#" className="btn-link toggle-sidebar d-lg-none pg pg-menu" data-toggle="sidebar">
        </a>
        <div>
          <div className="brand inline">
            <img src="img/logo-white.png" alt="logo" data-src="img/logo-white.png" height="36" />
          </div>
        </div>
        <div className="d-flex align-items-center">
          <div className="pull-left p-r-10 fs-14 font-heading d-lg-block d-none">
            <span className="active-timenodes">
              <i className="fa fa-sitemap"/>&nbsp;&nbsp;Active TimeNodes:&nbsp;
            </span>
            <span className="timenode-count">{this.props.keenStore.activeTimeNodes}</span>
          </div>
          <div className="left-separator pull-left p-l-10 fs-14 font-heading d-lg-block d-none">
            <span className="active-timenodes">
              <i className="fa fa-th-large" />&nbsp;Current Block Number:&nbsp;
            </span>
            <span className="timenode-count">{this.state.blocknumber}</span>
          </div>
        </div>
        <div className="d-flex">
          <div className="search-link d-lg-inline-block d-none" onClick={() => {this.props.updateSearchState(true);}}>
            <i className="pg-search"></i>
            Search by Address
          </div>
        </div>
      </div>
    );
  }
}

Header.propTypes = {
  updateSearchState: PropTypes.any,
  web3Service: PropTypes.any,
  keenStore: PropTypes.any
};

export default Header;