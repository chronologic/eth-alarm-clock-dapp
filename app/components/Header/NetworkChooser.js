import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { observer, inject } from 'mobx-react';
import { Networks, CUSTOM_PROVIDER_NET_ID } from '../../config/web3Config';
import ConfirmModal from '../Common/ConfirmModal';

const DEFAULT_NETWORK_ID = Networks[1].id; // Mainnet

@inject('web3Service')
@inject('timeNodeStore')
@observer
class NetworkChooser extends Component {
  constructor(props) {
    super(props);

    // We use two network providers
    // 1. Metamask injected provider for scheduling - This can be changed only when MetaMask is present
    // 2. Separate provider for TimeNodes if selected - This can be changed on platforms without MetaMask
    this.state = {
      metaMaskNetworkId: DEFAULT_NETWORK_ID,
      timeNodeNetworkId: this.checkSelectedProvider(),
      onTimeNodeScreen: props.onTimeNodeScreen,
      selectedNetId: null // Used for communicating the selected network to the modal
    };

    this._handleSelectedNetworkChange = this._handleSelectedNetworkChange.bind(this);
    this.changeProvider = this.changeProvider.bind(this);
  }

  async componentDidMount() {
    const { web3Service } = this.props;
    await web3Service.init();

    if (web3Service.network.id !== this.state.metaMaskNetworkId)
      this.setState({ metaMaskNetworkId: web3Service.network.id });

    if (!localStorage.getItem('selectedProviderId')) {
      if (web3Service.network.id !== this.state.timeNodeNetworkId)
        this.setState({ timeNodeNetworkId: web3Service.network.id });
    }
  }

  static getDerivedStateFromProps(newProps) {
    return {
      onTimeNodeScreen: newProps.onTimeNodeScreen
    };
  }

  /*
   * Checks if the TimeNode provider is different from the global
   * MetaMask provider.
   */
  checkSelectedProvider() {
    const { timeNodeStore } = this.props;

    const { id, endpoint } = timeNodeStore.getCustomProvider();

    if (id && endpoint) {
      timeNodeStore.customProviderUrl = endpoint;
      return id;
    }

    return DEFAULT_NETWORK_ID;
  }

  hasCustomProvider() {
    const selectedProviderId = parseInt(this.props.storageService.load('selectedProviderId'));

    return selectedProviderId === CUSTOM_PROVIDER_NET_ID;
  }

  _handleSelectedNetworkChange(event) {
    const selectedNetId = parseInt(event.target.value);
    const $ = window.jQuery;

    const isCustomSelected = selectedNetId === CUSTOM_PROVIDER_NET_ID;
    const modalToShow = isCustomSelected ? '#customProviderModal' : '#confirmProviderChangeModal';

    this.setState({ selectedNetId }); // Let the modal know which network was selected

    $(modalToShow).modal({
      show: true,
      backdrop: isCustomSelected // Backdrop breaks #confirmProviderChangeModal, so enable only for custom
    });
  }

  changeProvider() {
    // Retrieve the selected network id in the modal
    const { selectedNetId } = this.state;

    const selectedProviderEndpoint = Networks[selectedNetId].endpoint;
    this.props.timeNodeStore.setCustomProvider(selectedNetId, selectedProviderEndpoint);

    // Once we read the new network ID, reset it
    this.setState({
      selectedNetId: null
    });
  }

  componentWillUnmount() {
    clearInterval(this.interval);
  }

  render() {
    const { timeNodeNetworkId, metaMaskNetworkId } = this.state;
    const blockNumberString = blockNumber => (blockNumber ? ' at #' + blockNumber : '');

    if (!this.state.onTimeNodeScreen) {
      return (
        <span>
          {Networks[metaMaskNetworkId].name}
          {blockNumberString(this.props.web3Service.latestBlockNumber)}
        </span>
      );
    }

    // Return this only if the user enters the TimeNode screen.
    return (
      <span id="networkChooser">
        <span id="networkChooserSelect">
          <select
            className="form-control d-inline"
            value={timeNodeNetworkId}
            onChange={this._handleSelectedNetworkChange}
          >
            {Object.keys(Networks).map(index => {
              if (Networks[index].showInChooser) {
                return (
                  <option key={index} value={index}>
                    {Networks[index].name}
                  </option>
                );
              }
            })}
            <option key={CUSTOM_PROVIDER_NET_ID} value={CUSTOM_PROVIDER_NET_ID}>
              Custom...
            </option>
          </select>
        </span>
        {blockNumberString(this.props.timeNodeStore.providerBlockNumber)}

        <ConfirmModal
          modalName="confirmProviderChange"
          modalTitle="You are about to change your TimeNode provider."
          modalBody="Are you sure you want to change it? Your TimeNode will be stopped."
          onConfirm={this.changeProvider}
        />
      </span>
    );
  }
}

NetworkChooser.propTypes = {
  storageService: PropTypes.any,
  web3Service: PropTypes.any,
  timeNodeStore: PropTypes.any,
  onTimeNodeScreen: PropTypes.bool
};

export default NetworkChooser;
