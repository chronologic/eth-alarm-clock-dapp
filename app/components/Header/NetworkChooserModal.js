import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { observer, inject } from 'mobx-react';
import { Networks } from '../../config/web3Config';
import ConfirmModal from '../Common/ConfirmModal';

@inject('timeNodeStore')
@observer
class NetworkChooserModal extends Component {
  constructor(props) {
    super(props);

    this.changeProvider = this.changeProvider.bind(this);
  }

  changeProvider() {
    const selectedNetId = this.props.timeNodeStore.proposedNewNetId;

    const selectedProviderEndpoint = Networks[selectedNetId].endpoint;
    this.props.timeNodeStore.setCustomProvider(selectedNetId, selectedProviderEndpoint);

    // Once we read the new network ID, reset it
    this.props.timeNodeStore.proposedNewNetId = null;
  }

  render() {
    return (
      <ConfirmModal
        modalName="confirmProviderChange"
        modalTitle="You are about to change your TimeNode provider."
        modalBody="Are you sure you want to change it? Your TimeNode will be stopped."
        onConfirm={this.changeProvider}
      />
    );
  }
}

NetworkChooserModal.propTypes = {
  timeNodeStore: PropTypes.any
};

export default NetworkChooserModal;
