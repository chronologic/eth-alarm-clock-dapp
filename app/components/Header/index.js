export { Header } from './Header';
export { NetworkChooser } from './NetworkChooser';
