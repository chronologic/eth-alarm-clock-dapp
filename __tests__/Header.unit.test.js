import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'mobx-react';
import Web3Service from '../app/services/web3';
import Header from '../app/components/Header/index';

describe('Header', () => {
  it('displays current block number', () => {
    const CURRENT_BLOCK = 5361200;

    const web3Service = new Web3Service({
      web3: {
        eth: {
          getBlockNumber(callback) {
            callback(null, CURRENT_BLOCK);
          }
        }
      }
    });

    const keenStore = {};
    const eacService = {
      getActiveContracts: () => {
        return {};
      }
    };
    const featuresService = {};

    const injectables = {
      featuresService,
      keenStore,
      web3Service,
      eacService
    };

    let mockedRender = renderer.create(
      <Provider {...injectables}>
        <Header />
      </Provider>
    );

    let tree = mockedRender.toJSON();
    expect(tree).toMatchSnapshot();

    const blockNumberDisplay = tree.children
      .find(el => el.props.className === 'd-flex align-items-center')
      .children.find(
        el => el.props.className === 'pull-left p-r-10 fs-14 font-heading d-lg-block d-none'
      )
      .children.find(el => el.props.className === 'timenode-count');

    expect(blockNumberDisplay).toMatchSnapshot();
  });
});
