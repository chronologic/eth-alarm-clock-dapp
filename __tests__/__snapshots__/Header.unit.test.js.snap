// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header displays current block number 1`] = `
<div
  className="header"
>
  <a
    className="btn-link toggle-sidebar d-lg-none pg pg-menu"
    data-toggle="sidebar"
    href="#"
  />
  <div>
    <div
      className="brand inline"
    >
      <img
        alt="logo"
        data-src="img/logo-white.png"
        height="36"
        src="img/logo-white.png"
      />
    </div>
  </div>
  <div
    className="d-flex align-items-center"
  >
    <div
      className="pull-left p-r-10 fs-14 font-heading d-lg-block d-none"
    >
      <span
        className="active-timenodes"
      >
        <i
          className="fa fa-sitemap"
        />
          Active TimeNodes: 
      </span>
      <span
        className="timenode-count"
      />
    </div>
    <div
      className="left-separator pull-left p-l-10 fs-14 font-heading d-lg-block d-none"
    >
      <span
        className="active-timenodes"
      >
        <i
          className="fa fa-th-large"
        />
         Current Block Number: 
      </span>
      <span
        className="timenode-count"
      >
        5361200
      </span>
    </div>
  </div>
  <div
    className="d-flex"
  >
    <div
      className="search-link d-lg-inline-block d-none"
      onClick={[Function]}
    >
      <i
        className="pg-search"
      />
      Search by Address
    </div>
  </div>
</div>
`;

exports[`Header displays current block number 2`] = `
<span
  className="timenode-count"
/>
`;
